### PR TITLE
Cpf 4666/permission menu dao controller for table view create and detail view edit and delete

### PR DESCRIPTION
**Pull request by olhabn**
https://nanopay.atlassian.net/browse/CPF-4666

### CODE REVIEW

File: src/foam/comics/v2/DAOControllerConfig.js [hunk 1]
```diff
@@ -138,8 +138,11 @@ foam.CLASS({
     },
     {
       class: 'FObjectProperty',
-      of: 'foam.comics.v2.CRUDActionsAuth',
-      name: 'CRUDActionsAuth'
+      of: 'foam.comics.v2.CRUDActionsPredicate',
```
Comment (kgrgreer) [inline src/foam/comics/v2/DAOControllerConfig.js]: What advantage do we get by making this be a nested Object rather than just adding three Predicate properties here? Are we using the CRUDActionsPredicate somewhere else?

File: src/foam/lib/PermissionPredicate.js [hunk 2]
```diff
@@ -0,0 +1,47 @@
+foam.CLASS({
```
Comment (kgrgreer) [inline src/foam/lib/PermissionPredicate.js:7]: Add copyright header.

File: src/foam/comics/v2/CRUDActionsPredicate.js [hunk 3]
```diff
@@ -0,0 +1,18 @@
+foam.CLASS({
```
Comment (jlhughes) [inline src/foam/comics/v2/CRUDActionsPredicate.js]: license header, please - and any other files it's missing.

File: src/foam/comics/v2/CRUDActionsPredicate.js [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+  package: 'foam.comics.v2',
+  name: 'CRUDActionsPredicate',
+  properties: [
+    {
```
Comment (jlhughes) [inline src/foam/comics/v2/CRUDActionsPredicate.js]: why not 'read'? I may find further on but how do we hide actions for example?

File: src/foam/comics/v2/DAOBrowseControllerView.js [hunk 5]
```diff
@@ -98,12 +98,16 @@ foam.CLASS({
...
 
       this.addClass(this.myClass())
-      .add(this.slot(function(data, config, config$CRUDActionsAuth$create, config$browseBorder, config$browseViews, config$browseTitle) {
-        var createAction = config$CRUDActionsAuth$create
-            ? self.CREATE.clone().copyFrom({
-              availablePermissions: self.CREATE.availablePermissions.concat(config$CRUDActionsAuth$create)
-            })
-            : self.CREATE;
+      .add(this.slot(function(data, config, config$browseBorder, config$browseViews, config$browseTitle) {
+        if ( ! config.CRUDPermission || ! config.CRUDPermission.create || ! config.CRUDPermission.create.f ) {
```
Comment (jlhughes) [inline src/foam/comics/v2/DAOBrowseControllerView.js]: perhaps simplified with: 
`
  config.createEnabled = false;
  config.createEnabled = config.CRUDPermission && config.CRUDPermission.create && config.CRUDPermission.create.f ? config.CRUDPermission.f() || true; 
`

File: src/foam/comics/v2/DAOSummaryView.js [hunk 6]
```diff
@@ -150,30 +150,31 @@ foam.CLASS({
...
-            deleteAction = self.DELETE.clone().copyFrom({
-              availablePermissions: self.DELETE.availablePermissions.concat(deleteArray)
-            });
+        .add(self.slot(function(data, data$id, config$viewBorder, viewView) {
+
+          if ( ! self.config.CRUDPermission) {
+            self.config.editEnabled = true;
+            self.config.deleteEnabled = true;
+          } else {
+            if ( ! self.config.CRUDPermission.update || ! self.config.CRUDPermission.update.f ) {
```
Comment (jlhughes) [inline src/foam/comics/v2/DAOSummaryView.js]: Simplify if you can, similar to above.

File: src/foam/lib/PermissionPredicate.js [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+    }
+  ],
+  javaImports: [
+    'foam.nanos.auth.AuthService'
+  ],
+
+  methods: [
+    {
+      name: 'f',
+      javaCode: `
```
Comment (jlhughes) [inline src/foam/lib/PermissionPredicate.js:29]: The PermissionPredicates above, when is x passed in? 
Can we not pass 'x' as the 'obj' in f(obj)?  On the java side will this not be the system context?

File: src/foam/lib/PermissionPredicate.js (hunk 2 above)
Comment (jlhughes) [reply src/foam/lib/PermissionPredicate.js:7]: @olhabn the current header. 
`/**
 * @license
 * Copyright 2020 The FOAM Authors. All Rights Reserved.
 * http://www.apache.org/licenses/LICENSE-2.0
 */
`

File: src/foam/lib/PermissionPredicate.js (hunk 7 above)
Comment (olhabn) [reply src/foam/lib/PermissionPredicate.js:29]: Fixed

File: src/foam/comics/v2/DAOControllerConfig.js (hunk 1 above)
Comment (olhabn) [reply src/foam/comics/v2/DAOControllerConfig.js]: Fixed

**Review by jlhughes — APPROVED** [review]

File: src/foam/comics/v2/DAOControllerConfig.js [hunk 8]
```diff
@@ -138,8 +138,11 @@ foam.CLASS({
...
+      name: 'CRUDPermission',
```
Comment (jlhughes) [inline src/foam/comics/v2/DAOControllerConfig.js]: property names should not begin with upper case. 
```suggestion
      name: 'crudPermission',  
```

File: src/foam/comics/v2/DAOControllerConfig.js (hunk 1 above)
Comment (jlhughes) [reply src/foam/comics/v2/DAOControllerConfig.js]: Why do we have this permission and still have 'created/edit/deletedEnabled' above?

File: src/foam/comics/v2/DAOSummaryView.js [hunk 9]
```diff
@@ -150,30 +150,31 @@ foam.CLASS({
...
-
-            deleteArray = deleteArray.map((permission) => permission.replace('%', data$id));
-
-            deleteAction = self.DELETE.clone().copyFrom({
-              availablePermissions: self.DELETE.availablePermissions.concat(deleteArray)
-            });
+        .add(self.slot(function(data, data$id, config$viewBorder, viewView) {
+
+          if ( ! self.config.CRUDPermission) {
+            self.config.editEnabled = true;
```
Comment (jlhughes) [inline src/foam/comics/v2/DAOSummaryView.js]: If we are retaining editEnabled (and others above) for backwards compatibility (I presume), then self.config.editEnabled is either set or already has a default.

**Review by jlhughes — APPROVED** [review]

File: src/foam/comics/v2/DAOBrowseControllerView.js [hunk 10]
```diff
@@ -77,8 +77,12 @@ foam.CLASS({
   actions: [
     {
       name: 'create',
-      isAvailable: function(config$createEnabled) {
-        return config$createEnabled;
+      isAvailable: function(config) {
+        try {
+          return ! config.create || ! config.create.f ? true : config.create.f();
```
Comment (kgrgreer) [inline src/foam/comics/v2/DAOBrowseControllerView.js:82]: Why is this needed?

File: src/foam/comics/v2/DAOControllerConfig.js [hunk 11]
```diff
@@ -119,27 +119,16 @@ foam.CLASS({
       }
     },
     {
-      class: 'Boolean',
-      name: 'createEnabled',
-      documentation: 'If set to false, the "Create" button will not be visible.',
-      value: true
+      class: 'foam.mlang.predicate.PredicateProperty',
+      name: 'create'
```
Comment (kgrgreer) [inline src/foam/comics/v2/DAOControllerConfig.js:123]: Doesn't this cause a name conflict?

File: src/foam/comics/v2/DAOSummaryView.js [hunk 12]
```diff
@@ -107,8 +107,12 @@ foam.CLASS({
   actions: [
     {
       name: 'edit',
-      isAvailable: function(config$editEnabled) {
-        return config$editEnabled;
+      isAvailable: function(config) {
+        try {
+          return ! config.update || ! config.update.f  ? true : config.update.f();
```
Comment (kgrgreer) [inline src/foam/comics/v2/DAOSummaryView.js:112]: Why is this needed? Why can't you just always call config.update.f()?

File: src/foam/comics/v2/DAOSummaryView.js [hunk 13]
```diff
@@ -122,8 +126,12 @@ foam.CLASS({
     },
     {
       name: 'delete',
-      isAvailable: function(config$deleteEnabled) {
-        return config$deleteEnabled;
+      isAvailable: function(config) {
+        try {
+          return ! config.delete || ! config.delete.f ? true : config.delete.f();
```
Comment (kgrgreer) [inline src/foam/comics/v2/DAOSummaryView.js:131]: same

File: src/foam/lib/PermissionPredicate.js [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+  name: 'PermissionPredicate',
+  extends: 'foam.mlang.predicate.AbstractPredicate',
+  implements: ['foam.core.Serializable'],
+  imports: [
+    'auth',
+  ],
+  properties: [
+    {
+      name: 'args',
+      class: 'StringArray'
```
Comment (kgrgreer) [inline src/foam/lib/PermissionPredicate.js:18]: Better to have only one permission and then AND or OR them together if that's what you need. Otherwise, it isn't clear what it means when you have multiple permissions.

File: src/foam/lib/PermissionPredicate.js [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+    'foam.nanos.auth.AuthService'
+  ],
+
+  methods: [
+    {
+      name: 'f',
+      javaCode: `
+        X x = (X) obj;
+        AuthService auth = (AuthService) x.get("auth");
+        if ( args_ == null || args_.length <= 0 ) {
```
Comment (kgrgreer) [inline src/foam/lib/PermissionPredicate.js:32]: If no authservice, then we're probably running in a demo or outside of nanos, in which case I think we should avoid a NPE and still allow.

File: src/foam/lib/PermissionPredicate.js [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+          return true;
+        }
+
+        for ( String permission : args_ ) {
+          if ( auth.check(x, permission) )
+            return true;
+        }
+        return false;
+      `,
+      code: async function() {
```
Comment (kgrgreer) [inline src/foam/lib/PermissionPredicate.js:42]: Does this work? Don't predicates need to return synchronously?

File: src/foam/comics/v2/DAOSummaryView.js (hunk 12 above)
Comment (jlhughes) [reply src/foam/comics/v2/DAOSummaryView.js:112]: yes, this was before each method was changed to a Predicate.  And the issue is no default value of a TRUE predicate. 
return config.update.f()

File: src/foam/comics/v2/DAOControllerConfig.js (hunk 11 above)
Comment (jlhughes) [reply src/foam/comics/v2/DAOControllerConfig.js:123]: @olhabn revert to create/edit/deletedEnabled. 
and add default Predicate of TRUE.

File: src/foam/lib/PermissionPredicate.js (hunk 16 above)
Comment (jlhughes) [reply src/foam/lib/PermissionPredicate.js:42]: so need to use slots, default to false, then update when the call returns.

Comment (olhabn) [reply src/foam/lib/PermissionPredicate.js:42]: > Does this work? Don't predicates need to return synchronously?

I believe it does. I'm not sure where I can find information about that

Comment (olhabn) [reply src/foam/lib/PermissionPredicate.js:42]: > so need to use slots, default to false, then update when the call returns.

But other predicates return boolean

File: src/foam/lib/PermissionPredicate.js (hunk 15 above)
Comment (olhabn) [reply src/foam/lib/PermissionPredicate.js:32]: Could you please tell me what is NPE?

File: src/foam/lib/PermissionPredicate.js (hunk 14 above)
Comment (olhabn) [reply src/foam/lib/PermissionPredicate.js:18]: fixed https://github.com/foam-framework/foam2/pull/3234

File: src/foam/comics/v2/DAOSummaryView.js (hunk 13 above)
Comment (olhabn) [reply src/foam/comics/v2/DAOSummaryView.js:131]: fixed https://github.com/foam-framework/foam2/pull/3234

File: src/foam/comics/v2/DAOSummaryView.js (hunk 12 above)
Comment (olhabn) [reply src/foam/comics/v2/DAOSummaryView.js:112]: fixed https://github.com/foam-framework/foam2/pull/3234

File: src/foam/comics/v2/DAOControllerConfig.js (hunk 11 above)
Comment (olhabn) [reply src/foam/comics/v2/DAOControllerConfig.js:123]: fixed https://github.com/foam-framework/foam2/pull/3234

File: src/foam/comics/v2/DAOBrowseControllerView.js (hunk 10 above)
Comment (olhabn) [reply src/foam/comics/v2/DAOBrowseControllerView.js:82]: fixed https://github.com/foam-framework/foam2/pull/3234
